src/dscanner/reports.d

```diff
@@ -0,0 +1,160 @@
+//			Copyright Brian Schott (Hackerpilot) 2012.
+// Distributed under the Boost Software License, Version 1.0.
+//	  (See accompanying file LICENSE_1_0.txt or copy at
+//			http://www.boost.org/LICENSE_1_0.txt)
+
+module dscanner.reports;
+
+import std.json;
+import std.algorithm : map;
+import std.array : split, array;
+
+import dscanner.analysis.base : Message, MessageSet;
+
+class SonarQubeGenericIssueDataReporter
+{
+	enum Type
+	{
+		bug = "BUG",
+		vulnerability = "VULNERABILITY",
+		codeSmell = "CODE_SMELL"
+	}
+
+	enum Severity
+	{
+		blocker = "BLOCKER",
+		critical = "CRITICAL",
+		major = "MAJOR",
+		minor = "MINOR",
+		info = "INFO"
+	}
+
+	struct Issue
+	{
+		string engineId;
+		string ruleId;
+		Location primaryLocation;
+		string type;
+		string severity;
+		int effortMinutes;
+		Location[] secondaryLocations;
+	}
+
+	struct Location
+	{
+		string message;
+		string filePath;
+		TextRange textRange;
+	}
+
+	struct TextRange
+	{
+		// SonarQube Generic Issue only allows specifying start line only
+		// or the complete range, for which start and end has to differ
+		// D-Scanner does not provide the complete range info
+		long startLine;
+	}
+
+	private Issue[] _issues;
+
+	void addMessageSet(MessageSet messageSet)
+	{
+		_issues ~= toIssues(messageSet);
+	}
+	
+	void addMessage(Message message, bool isError = false)
+	{
+		_issues ~= toIssue(message, isError);
+	}
+
+	string getContent()
+	{
+		JSONValue result = [
+			"issues" : JSONValue(_issues.map!(e => toJson(e)).array)
+		];
+		return result.toString();
+	}
+
+	private JSONValue toJson(Issue issue)
+	{
+		// dfmt off
+		return JSONValue([
+			"engineId": JSONValue(issue.engineId),
+			"ruleId": JSONValue(issue.ruleId),
+			"severity": JSONValue(issue.severity),
+			"type": JSONValue(issue.type),
+			"primaryLocation": JSONValue([
+				"message": JSONValue(issue.primaryLocation.message),
+				"filePath": JSONValue(issue.primaryLocation.filePath),
+				"textRange": JSONValue([
+					"startLine": JSONValue(issue.primaryLocation.textRange.startLine)
+				]),
+			]),
+		]);
+		// dfmt on
+	}
+
+	private Issue[] toIssues(MessageSet messageSet)
+	{
+		return messageSet[].map!(e => toIssue(e)).array;
+	}
+
+	private Issue toIssue(Message message, bool isError = false)
+	{		
+		// dfmt off
+		Issue issue = {
+			engineId: "dscanner",
+			ruleId : message.key,
+			severity : (isError) ? Severity.blocker : getSeverity(message.key),
+			type : getType(message.key),
+			primaryLocation : getPrimaryLocation(message)
+		};
+		// dfmt on
+		return issue;
+	}
+
+	private Location getPrimaryLocation(Message message)
+	{
+		return Location(message.message, message.fileName, TextRange(message.line));
+	}
+
+	private string getSeverity(string key)
```

Comment:
could this be made `public static` or possibly be moved outside the class? I think this might be useful in other applications like editors directly integrating DScanner. It should probably be documented how the naming affects the severity then too.

I think the method could also take in a `const ref Message message` argument so in case we add some severity/type enum to the message, it can simply translate to the SonarQube enums and otherwise now just only use `.key`.